File: everything_parsing/28-grammar-tool-project2/java/Main.java
import java.nio.file.Files;
import java.nio.file.Path;

public final class Main {

    static final String[] FILES = {
        "fixtures/stmts.grammar",
        "fixtures/expr.grammar",
        "fixtures/json.grammar",
    };

    public static void main(String[] args) throws Exception {
        StringBuilder sb = new StringBuilder();
        sb.append("=== Module 28 - Grammar Analysis Tool (CSE 340 Project 2) ===\n\n");
        sb.append("one program, seven sections, each a component from Modules 20-27:\n");
        sb.append("symbols | nullable | FIRST | FOLLOW | left-recursion-free | left-factored | LL(1)\n\n");

        for (String file : FILES) {
            Grammar g = Grammar.parse(Files.readString(Path.of(file)));
            GrammarTool tool = new GrammarTool(g);

            sb.append("##################### ").append(file).append(" #####################\n\n");
            sb.append("input grammar:\n");
            for (Grammar.Production p : g.productions)
                sb.append("  ").append(p).append('\n');
            sb.append('\n');
            sb.append(tool.report());
            sb.append('\n');
        }

        System.out.print(sb.toString().replace("\r\n", "\n"));
    }
}

File: everything_parsing/28-grammar-tool-project2/js/main.mjs
// node js/main.mjs   — identical output bytes to the Java build.

import { readFileSync } from "node:fs";
import { Grammar } from "./grammar.mjs";
import { GrammarTool } from "./grammartool.mjs";

const FILES = [
  "fixtures/stmts.grammar",
  "fixtures/expr.grammar",
  "fixtures/json.grammar",
];

let sb = "";
const p = (s = "") => { sb += s + "\n"; };

p("=== Module 28 - Grammar Analysis Tool (CSE 340 Project 2) ===");
p();
p("one program, seven sections, each a component from Modules 20-27:");
p("symbols | nullable | FIRST | FOLLOW | left-recursion-free | left-factored | LL(1)");
p();

for (const file of FILES) {
  const g = Grammar.parse(readFileSync(file, "utf8"));
  const tool = new GrammarTool(g);

  p("##################### " + file + " #####################");
  p();
  p("input grammar:");
  for (const prod of g.productions)
    p("  " + prod.lhs + " -> " + (prod.rhs.length === 0 ? "epsilon" : prod.rhs.join(" ")));
  p();
  sb += tool.report();
  p();
}

process.stdout.write(sb);
